### PR TITLE
Inform the user about a missing `fabric.backward()` call

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

The following works without issues and is technically correct:

```py
fabric = Fabric(accelerator="cpu", devices=1)

module = torch.nn.Linear(2, 2)
module = fabric.setup(module)

loss = module(torch.rand(2, 2))
loss.sum().backward()
```
A call to `fabric.backward()` is not necessary because in case of single-device, full precision execution, `fabric.backward(loss)` and `loss.backward()` are fully equivalent.

However, if you switch to e.g. `precision="bf16-mixed"`, `strategy="deepspeed"`, `strategy="fsdp"`, etc., this would silently be incorrect. In `bf16-mixed` for example, it would silently not unscale the loss/gradient, leading to issues later on.

This PR adds a backward hook on the module passed to `Fabric.setup()`, such that when `fabric.backard()` is not called, will raise an error and inform the user.

This is **not a breaking change** because:
- For strategies and precision settings that don't have a custom backward, you can still write `loss.backward()` without issues. No errors will be raised.
- If you were using strategies or precision settings before that have custom backward implemented, and you were not calling `fabric.backward()` you will see an error now but it means you previously had a silent bug in your code.

There was an alternative in #18844 that would have removed the need to call `fabric.backward()` altogether, but we decided it is too soon to land it.

----
📚 Documentation preview 📚: https://pytorch-lightning--19447.org.readthedocs.build/en/19447/

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

File: src/lightning/fabric/fabric.py [hunk 1]
```diff
@@ -142,6 +142,7 @@ def __init__(
         self._loggers = loggers if isinstance(loggers, list) else [loggers]
         self._models_setup: int = 0
         self._launched: bool = False
+        self._backward_called: bool = False
```
Comment (carmocca) [inline src/lightning/fabric/fabric.py:145]: Shouldn't this be in the `_FabricModule`? This wouldn't work well with multiple models

Comment (awaelchli) [reply src/lightning/fabric/fabric.py:145]: What would be an example of it not working with multiple models? I have tried to put it into the FabricModule first, but it was more complex and I want to avoid it if not needed.

Comment (carmocca) [reply src/lightning/fabric/fabric.py:145]: I mean if you wrap two models and call `fabric.backward()` with the loss associated to one but `loss.backward()` with the other. It will work if they run sequentially of course but I find it mildly worrying to share module-specific attributes in a different object (Fabric).

What's complex about it? You always have the module reference available no?

Comment (awaelchli) [reply src/lightning/fabric/fabric.py:145]: `fabric.backward()` doesn't have the module reference available in the general case. You were probably thinking that `fabric.backward()` would be able to do `model._backward_called = True` and then the backward hooks would check this on the model instance, but that's not possible if `model=None`, regardless of one or multiple models.

I just added two more scenarios with multiple models to the test case. Please take a look at them. At the moment, can't find any edge cases that would be problematic.

File: tests/tests_fabric/test_fabric.py [hunk 2]
```diff
@@ -610,12 +610,68 @@ def test_rank_properties():
...
+    ("auto", "32-true", False),
+    ("auto", "bf16-true", False),
+    ("auto", "bf16-mixed", True),
+    pytest.param("fsdp", "32-true", True, marks=RunIf(min_cuda_gpus=1, min_torch="2.0.0")),
+])
+@pytest.mark.parametrize("setup_method", ["setup", "setup_module"])
+@mock.patch("lightning.fabric.accelerators.mps.MPSAccelerator.is_available", return_value=False)
+def test_backward_required(_, strategy, precision, error_expected, setup_method):
+    """Test under which strategy and precision configurations the `fabric.backward()` call is required."""
+    fabric = Fabric(strategy=strategy, precision=precision, devices=1)
```
Comment (carmocca) [inline tests/tests_fabric/test_fabric.py:629]: This test fails on single-device CUDA machines

File: src/lightning/fabric/fabric.py (hunk 1 above)
Comment (carmocca) [reply src/lightning/fabric/fabric.py:145]: You could register a forward hook that sets it to True to the model

**Review by carmocca — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Comment by carmocca** [comment]
This doesn't work for models that don't return a Tensor or a tuple of tensors:

```python
import torch
import lightning as L

class MyModel(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.l = torch.nn.Linear(1, 1)

    def forward(self, x):
        y = self.l(x)
        return [y, "foo"]

model = MyModel()
fabric = L.Fabric(precision="16-mixed")
model = fabric.setup(model)
x = torch.randn(2, 1, device=fabric.device)
y, foo = model(x)
loss = y.sum()
print(fabric._backward_called)
loss.backward()
print(fabric._backward_called)
```

This warning appears:
https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/module.py#L1597-L1601

We should avoid this warning since returning dicts or dataclasses is not rare

**Comment by awaelchli** [comment]
I didnt know this. Then we need to revert this PR, this approach will never work for this.

**Comment by carmocca** [comment]
I also didnt know. If we avoid the warning we could keep this for the models that return tensors. The hook wont run for the others but it shouldnt cause any issues

**Comment by awaelchli** [comment]
I don't know if this is overall the better approach, but we could alternatively register a backward-hook on any tensor returned by the `_FabricModule` (even if contained in a dict or nested structure): #19493
